_posts/2017-12-08-totw-123.md

```diff
@@ -0,0 +1,132 @@
+
+## How to store values
+
+This tip discusses several ways of storing values. Here we use class member
+variables as an example, but many of the points below also apply to local
+variables.
+
+```c++
+#include <memory>
+#include "third_party/absl/types/optional.h"
+#include ".../bar.h"
+
+class Foo {
+  ...
+ private:
+  Bar val_;
+  absl::optional<Bar> opt_;
+  std::unique_ptr<Bar> ptr_;
+};
+```
+
+### As a bare object
+
+This is the simplest way. `val_` is constructed and destroyed at the beginning
+of `Foo`'s constructor and at the end of `Foo`'s destructor, respectively. If
+`Bar` has a default constructor, it doesn't even need to be initialized
+explicitly.
+
+`val_` is very safe to use, because its value can't be null. This removes a
+class of potential bugs.
+
+But bare objects are not very flexible:
+
+*   The lifetime of `val_` is fundamentally tied to the lifetime of its parent
+    `Foo` object, which is sometimes not desirable. If `Bar` supports move or
+    swap operations, the contents of `val_` can be replaced using these
+    operations, while any existing pointers or references to `val_` continue
+    pointing or referring to the same `val_` object (as a container), not to the
+    value stored in it.
+*   Any arguments that need to be passed to `Bar`'s constructor need to be
+    computed inside the initializer list of `Foo`'s constructor, which can be
+    difficult if complicated expressions are involved.
+
+### As `absl::optional<Bar>`
+
+This is a good middle ground between simplicity of bare objects and flexibility
+of `std::unique_ptr`. The object is stored inside `Foo` but, unlike bare
+objects, `absl::optional` can be empty. It can be populated at any time by
+assignment (`opt_ = ...`) or by constructing the object in place
+(`opt_.emplace(...)`).
+
+Because the object is stored inline, the usual caveats about allocating large
+objects on the stack apply, just like for a bare object. Also be aware that an
+empty `absl::optional` uses as much memory as a populated one.
+
+Compared to a bare object, `absl::optional` has a few downsides:
+
+*   It's less obvious for the reader where object construction and destruction
+    occur.
+*   There is a risk of accessing an object which does not exist.
+
+### As `std::unique_ptr<Bar>`
+
+This is the most flexible way. The object is stored outside of `Foo`. Just like
+`absl::optional`, a `std::unique_ptr` can be empty. However, unlike
+`absl::optional`, it is possible to transfer ownership of the object to
+something else (through a move operation), to take ownership of the object from
+something else (at construction or through assignment), or to assume ownership
+of a raw pointer (at construction or through `ptr_ = absl::WrapUnique(...)`, see
+[TotW 126](/tips/126).
+
+When `std::unique_ptr` is null, it doesn't have the object allocated, and
+consumes only the size of a pointer[^deleter].
+
+This is the only viable solution if the object needs to outlive the scope of the
```

Comment:
"This is" -> Wrapping an object within a `std::unique_ptr`

(Otherwise, it seems to be saying that a unique_ptr is the only viable solution for outliving a unique_ptr ... :)

I'd also change "needs" to "may need"

Reply:
Done with adjustments, ptal.